### PR TITLE
Support for skew symmetric variables

### LINKED ISSUE #2146: Support for skew-symmetric matrix variables
I contribute to [PowerModelsDistribution](https://github.com/lanl-ansi/PowerModelsDistribution.jl). Ideally, we would like to use JuMP's support for matrix variables with symmetry more broadly. 

As we start from a complex-value problem with Hermitian matrix variables, we would love to see support for both symmetric (real part) and skew-symmetric (imaginary part) matrix variables. We have a work-around, but the printing of the variables looks different from other variables.

We can do something like the following to make sets of full and symmetric matrix variables:
```
Dict(i => JuMP.@variable(pm.model,
            [c in 1:N, d in 1:N], base_name="P_$(i)",
        ) for i in ids  )
Dict(i => JuMP.@variable(pm.model,
            [c in 1:N, d in 1:N], base_name="Wr_$(i)", Symmetric
        ) for i in ids  )
```
but the following is not supported:
```
Dict(i => JuMP.@variable(pm.model,
            [c in 1:N, d in 1:N], base_name="Wi_$(i)", SkewSymmetric
        ) for i in ids  )
```
Ideally, this would have (support for) zeros on the diagonal as well.

After some preliminary digging, it seems Julia does not have skew symmetric matrix variable primitives (https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/#Special-matrices-1, somewhat related discussion in https://github.com/JuliaLang/julia/issues/21224)

### PR DESCRIPTION
**Pull request by frederikgeth**
This is an attempt to support skew symmetric variables (https://github.com/JuliaOpt/JuMP.jl/issues/2146).

Note that the return type (affine expression) is different from the input type (variable ref) [here]( https://github.com/frederikgeth/JuMP.jl/blob/22fc8d29e549d7fee9666fc4d925b8eb03be785c/src/sd.jl#L113).

Keen to hear your feedback!

Closes #2146

### CODE REVIEW

File: src/sd.jl [hunk 1]
```diff
@@ -167,6 +202,25 @@ function build_variable(_error::Function, variables::Matrix{<:ScalarVariable}, :
...
+constrained after their creation.
+
+This function is used by the [`@variable`](@ref) macro as follows:
+```julia
+@variable(model, Q[1:2, 1:2] in SkewSymMatrixSpace())
+```
+"""
+function build_variable(_error::Function, variables::Matrix{<:ScalarVariable}, ::SkewSymMatrixSpace)
+    n = _square_side(_error, variables)
+    set = MOI.Reals((n^2-n)/2)
```
Comment (blegat) [inline src/sd.jl]: It should be `div(n^2 - n, 2)`, `n^2 - n` is even for any `n`.

File: src/sd.jl [hunk 2]
```diff
@@ -96,6 +97,40 @@ function vectorize(matrix::Matrix, ::SymmetricMatrixShape)
...
+Shape object for a skew symmetric square matrix of `side_dimension` rows and columns.
+The vectorized form contains the entries of the upper-right triangular part of
+the matrix (without the diagonal) given column by column (or equivalently, the
+entries of the lower-left triangular part given row by row). The diagonal is zero.
+"""
+struct SkewSymmetricMatrixShape <: AbstractShape
+    side_dimension::Int
+end
+function reshape_vector(vectorized_form::Vector{T}, shape::SkewSymmetricMatrixShape) where T
+    matrix = Matrix{JuMP.GenericAffExpr{Float64,JuMP.VariableRef}}(undef, shape.side_dimension, shape.side_dimension)
```
Comment (blegat) [inline src/sd.jl]: It should be `Matrix{Base.promote_type(T, MA.promote_operation(-, T), MA.promote_operation(zero, T))}` so that it works for any `T`

File: src/sd.jl [hunk 3]
```diff
@@ -96,6 +97,40 @@ function vectorize(matrix::Matrix, ::SymmetricMatrixShape)
...
+    k = 0
+    for j in 1:shape.side_dimension
+        for i in 1:j-1
+            k += 1
+            matrix[i, j] =  vectorized_form[k]
+            matrix[j, i] = -vectorized_form[k]
+        end
+    end
+    for j in 1:shape.side_dimension
+        matrix[j, j] =  0
```
Comment (blegat) [inline src/sd.jl]: Replace `0` by `zero(T)`

File: src/sd.jl (hunk 1 above)
Comment (frederikgeth) [reply src/sd.jl]: ok!

File: src/sd.jl (hunk 3 above)
Comment (frederikgeth) [reply src/sd.jl]: ok

File: src/sd.jl (hunk 2 above)
Comment (frederikgeth) [reply src/sd.jl]: thanks!

File: src/sd.jl [hunk 4]
```diff
@@ -96,6 +97,40 @@ function vectorize(matrix::Matrix, ::SymmetricMatrixShape)
...
+Shape object for a skew symmetric square matrix of `side_dimension` rows and columns.
+The vectorized form contains the entries of the upper-right triangular part of
+the matrix (without the diagonal) given column by column (or equivalently, the
+entries of the lower-left triangular part given row by row). The diagonal is zero.
+"""
+struct SkewSymmetricMatrixShape <: AbstractShape
+    side_dimension::Int
+end
+function reshape_vector(vectorized_form::Vector{T}, shape::SkewSymmetricMatrixShape) where T
+    matrix = Matrix{Base.promote_type(T, _MA.promote_operation(-, T), _MA.promote_operation(zero, T))}(undef, shape.side_dimension, shape.side_dimension)
```
Comment (blegat) [inline src/sd.jl]: This breaks 80-char limit

File: src/sd.jl [hunk 5]
```diff
@@ -96,6 +97,40 @@ function vectorize(matrix::Matrix, ::SymmetricMatrixShape)
...
+    k = 0
+    for j in 1:shape.side_dimension
+        for i in 1:j-1
+            k += 1
+            matrix[i, j] =  vectorized_form[k]
+            matrix[j, i] = -vectorized_form[k]
+        end
+    end
+    for j in 1:shape.side_dimension
+        matrix[j, j] =  zero(T)
```
Comment (blegat) [inline src/sd.jl]: Remove space before `zero`

File: src/sd.jl [hunk 6]
```diff
@@ -167,6 +202,25 @@ function build_variable(_error::Function, variables::Matrix{<:ScalarVariable}, :
     return VariablesConstrainedOnCreation(_vectorize_variables(_error, variables), set, shape)
 end
 
+"""
+    build_constraint(_error::Function, variables, ::SkewSymMatrixSpace)
+
+Return a `VariablesConstrainedOnCreation` of shape [`SkewSymMatrixSpace`](@ref)
```
Comment (blegat) [inline src/sd.jl]: `Space` ->`Shape`

File: test/variable.jl [hunk 7]
```diff
@@ -467,6 +467,19 @@ function test_variable_symmetric(ModelType)
...
+    model = ModelType()
+
+    @variable model x[1:2, 1:2] in SkewSymMatrixSpace()
+    @test x[1, 2] == -x[2, 1]
+    @test x[1, 1] == 0
+    @test x[2, 2] == 0
+    @test model[:x] === x
+
+    y = @variable model [1:2, 1:2] in SkewSymMatrixSpace()
+    @test y[1, 2] == -y[2, 1]
```
Comment (blegat) [inline test/variable.jl]: I would add a 3 x 3 example as the 2 x 2 examples only has 1 non-diag entry so it may not cover all corner cases

**Comment by frederikgeth** [comment]
Thanks, @blegat , I will update the pr and add the docs tomorrow

**Comment by frederikgeth** [comment]
So tests fail for `test_variable_skewsymmetric({JuMPExtension.MyModel})` for 3x3 matrices. This is the error:
```
  Got exception outside of a @test
  DimensionMismatch("arrays could not be broadcast to a common size; got a dimension with lengths 6 and 3")
  Stacktrace:
   [1] add_variable(::Main.JuMPExtension.MyModel, ::VariablesConstrainedOnCreation{MathOptInterface.Reals,SkewSymmetricMatrixShape,ScalarVariable{Float64,Float64,Float64,Float64}}, ::Array{String,2}) at ./broadcast.jl:490
```

Not sure where to go with this?

File: test/variable.jl [hunk 8]
```diff
@@ -467,6 +467,30 @@ function test_variable_symmetric(ModelType)
...
 end
 
+function test_variable_skewsymmetric(ModelType)
+    model = ModelType()
+
+    @variable model x[1:2, 1:2] in SkewSymMatrixSpace()
+    @test x[1, 2] == -x[2, 1]
+    @test iszero(x[1, 1])
+    @test iszero(x[2, 2])
+    @test model[:x] === x
```
Comment (blegat) [inline test/variable.jl]: Add test for the number of variables actually added to the inner model

File: src/sd.jl [hunk 9]
```diff
@@ -167,6 +203,25 @@ function build_variable(_error::Function, variables::Matrix{<:ScalarVariable}, :
...
+This function is used by the [`@variable`](@ref) macro as follows:
+```julia
+@variable(model, Q[1:2, 1:2] in SkewSymMatrixSpace())
+```
+"""
+function build_variable(_error::Function, variables::Matrix{<:ScalarVariable}, ::SkewSymMatrixSpace)
+    n = _square_side(_error, variables)
+    set = MOI.Reals(div(n^2 - n, 2))
+    shape = SkewSymmetricMatrixShape(n)
+    return VariablesConstrainedOnCreation(_vectorize_variables(_error, variables), set, shape)
```
Comment (blegat) [inline src/sd.jl]: `vectorize_variables` also include the diagonal, this why the test are failing. I would guess because of this, you add to many variables so test the number of variables added

File: test/variable.jl (hunk 8 above)
Comment (frederikgeth) [reply test/variable.jl]: I'm not sure how to access those?

File: src/sd.jl (hunk 9 above)
Comment (frederikgeth) [reply src/sd.jl]: ok

File: src/sd.jl (hunk 4 above)
Comment (frederikgeth) [reply src/sd.jl]: ok

File: src/sd.jl [hunk 10]
```diff
@@ -167,6 +205,25 @@ function build_variable(_error::Function, variables::Matrix{<:ScalarVariable}, :
...
+This function is used by the [`@variable`](@ref) macro as follows:
+```julia
+@variable(model, Q[1:2, 1:2] in SkewSymMatrixSpace())
+```
+"""
+function build_variable(_error::Function, variables::Matrix{<:ScalarVariable}, ::SkewSymMatrixSpace)
+    n = _square_side(_error, variables)
+    set = MOI.Reals(div(n^2 - n, 2))
+    shape = SkewSymmetricMatrixShape(n)
+    return VariablesConstrainedOnCreation(vectorize(variables, SkewSymmetricMatrixShape(n)), set, shape)
```
Comment (blegat) [inline src/sd.jl]: Here, you will ignore `variables` outside the upper triangle, add a TODO for throwing error in case the user set something that does not suit the skew symmetry like we do in `_vectorize_variables`

File: test/variable.jl (hunk 8 above)
Comment (blegat) [reply test/variable.jl]: https://www.juliaopt.org/JuMP.jl/dev/variables/#JuMP.num_variables

**Comment by martincornejo** [comment]
Any updates on this PR? This feature could be convenient for me

**Comment by odow** [comment]
Moved to https://github.com/jump-dev/JuMP.jl/pull/2416
